### ISSUE TITLE
Error: One or More Attributes Could Not Be Set

### ISSUE THREAD

**Issue by rrfranks** [SEPIA-Framework/sepia-docs#174]
Hi! I am trying to set up Sepia with Openhab on a raspberry pi.  I have had Openhab installed for a long time  on my pi, but recently decided recently to install Sepia on the same pi so I could have voice control over Openhab.  I successfully installed Sepia and am able to see my Openhab items in card format under the Smart Home option of Sepia .  Actually I can also turn lights on & off using the power button in Sepia's Smart Home.  However, when I choose a card to enter the Type or Room (through the drop down menu)  I get the following error:  

{
"result": "fail",
"error": "One or more attributes could not be set
}

Does anyone have any suggestions on what I am doing wrong?  Thank you for considering my question.

**Comment by fquirin** [comment]
Hi,

can you check your SEPIA Assist-Server log file at `SEPIA/sepia-assist-server/log.out` and look for entries like this `OpenHAB - 'writeDeviceAttribute' FAILED ...` please?
My first guess would be that OpenHAB refuses to write the data into the `tags` property (I think it was the tags field).

**Comment by fquirin** [comment]
Some additional log info submitted by @SHU-red:
```
2022-03-31 15:20:47 ERROR - OpenHAB - 'writeDeviceAttribute' FAILED - Device: dLedStrpBedRoom_Swt - Response: {"code":405,"error":"java.lang.NullPointerException","HTTP_REST_SUCCESS":false}
```

Error code 405 suggest that the openHAB server does not permit the write request.
I wonder if something changed in OH3? Are there any configurable permissions for the REST API?

**Comment by fquirin** [comment]
The SEPIA server will try to do a HTTP PUT request like this:
```
curl -X PUT --header "Content-Type: text/plain" --header "Accept: application/json" -d "" "http://{openHAB_IP}:8080/rest/items/{device_id}/tags/sepia-room%3Dlivingroom"
```

I'm not sure anymore if this was different in OH2, maybe it has to be changed to this:
```
curl -X PUT --header "Content-Type: text/plain" --header "Accept: application/json" -d "sepia-room=livingroom" "http://{openHAB_IP}:8080/rest/items/{device_id}/tags"
```

I think I have to install OH3 to test it ^^

**Comment by SHU-red** [comment]
Sorry for the lade reply!
I am absolutely willing to try out commands for you and provide the result (Although i do not know what im doing ; -) )

I just want to make some things clear regarding my setup:
1. As already mentioned i am running OH3 and Sepia and Sepia-TTS now in docker on my home server
2. I got the connection working thanks to your documentation via setting these variables as admin
2a. smarthome_hub_host = "http://192.myserverip:8080"
2b. smarthome_hub_name = "openhab"
2c. smarthome_hub_auth_type = "Basic"
2d. smarthome_hub_auth_data = put my "[username]:[password]" into a converter

I can see all OH3 items and their name
The only thing i can not do is set a Type and Room from within Sepia

I have not configured a semantic model in OH3 (which i think was not present in OH2 ???)
I do not know if Type and Room is something i also should be able to configure in OH?
Or if this is only Sepia-specific?

So basically i dont know if i did something wrong, if sepia has a bug or if i did not fully configure openhab.
Maybe i can proceed tomorrow searching for the root-cause and if you have any requests for me i will try to do my best to provide you the result

Thanks for your work

**Comment by SHU-red** [comment]
Oh man you are putting so much work into this ☺️

I also have a guess:
I defined all of my items via *.items file/script and not GUI.
In this case the chance is high, that items can not be changed via API.

I will test
1. Try to set type/room through sepia for a gui-generated item
2. Try to add sepia metadata via modifying *.items file

Can you give me a hint what metadata exactly to set?
What you saw there was my generic naming of my items. Can you tell me how I can set a "nice name" for using it for commands like "Ambient Light" instead of "dLedStripe_Swt"

Thanks again

**Comment by rrfranks** [comment]
Hi Everyone!  I am sorry for taking so long to reply.  Work has been very hectic this week and I am just now getting an opportunity to work on this again.  Let me see if I can assimilate all the suggestions and see where that leads me.  I can not thank you all enough for the effort you have put in to this...you all are GREAT!

**Comment by rrfranks** [comment]
Wow!  All your advice really paid off.  I added...

- the username/password info for OH to the Sepia config file 
- sepia tags to my OH item and 
- Porcupine Keyword Spotter addon to OH

and just like magic I was able to turn off & on a lamp with my voice!   Thank you all so much for giving so much time to help me out with this!!!!

**Comment by SHU-red** [comment]
@rrfranks 
Im very glad you got this working!

Porcumine sounds very interesting and could exactly be what i need!
Could you maybe add some information about your setup [here?](https://github.com/SEPIA-Framework/sepia-docs/discussions/177)

I especially have no idea how to create/use SmartHome Speakers with sepia openHAB and Google Assistant

**Comment by fquirin** [comment]
> Porcupine sounds very interesting and could exactly be what i need!

Porcupine is also part f the SEPIA client and [can be fully customized](https://github.com/SEPIA-Framework/sepia-html-client-app/blob/master/www/xtensions/picovoice/README.md) since Porcupine version v2.0 (with free access key) :slightly_smiling_face: . I'm currently implementing the v2.1 update to be in sync with their online console again.

> Can you give me a hint what metadata exactly to set?

[Here is a list](https://github.com/SEPIA-Framework/sepia-docs/wiki/Smart-Home-Controls#option-b-if-you-create-items-via-items-file-credits-to-mtothekay) of the SEPIA tags that you could set manually in OH.
If you look at your item using the REST API URL (simply replace `/settings/` with `/rest/`) in your browser you should see something like this:
```
{
    "link": "http://localhost:8080/rest/items/Huewhitelamp1_Brightness",
    "state": "0",
    "stateDescription": {
        "minimum": 0,
        "maximum": 100,
        "pattern": "%d %%",
        "readOnly": false,
        "options": []
    },
    "editable": true,
    "type": "Dimmer",
    "name": "Huewhitelamp1_Brightness",
    "label": "Brightness",
    "category": "Light",
    "tags": ["sepia-room\u003dlivingroom", "Control", "Light", "sepia-name\u003dLampe (1)", "sepia-type\u003dlight"],
    "groupNames": []
}
```

The tags might show up either as `"sepia-room\u003dlivingroom"` or `"sepia-room=livingroom"` that is just some encoding stuff.

> I defined all of my items via *.items file/script and not GUI
In this case the chance is high, that items can not be changed via API.

Ah yes. I never tried this but people mentioned that items are probably write protected in this case :grimacing: . Maybe I should explicitly mention it again in the openHAB docs section.

**Comment by fquirin** [comment]
Another update:

I just double-checked, if you use API token authentication you can use the `Bearer` auth. type and simply add your API token instead of base64(username:password) :wink: 
I've updated the docs accordingly.

**Comment by SHU-red** [comment]
> Oh man you are putting so much work into this relaxed
> 
> I also have a guess: I defined all of my items via *.items file/script and not GUI. In this case the chance is high, that items can not be changed via API.
> 
> I will test
> 
> 1. Try to set type/room through sepia for a gui-generated item
> 2. Try to add sepia metadata via modifying *.items file
> 
> Can you give me a hint what metadata exactly to set? What you saw there was my generic naming of my items. Can you tell me how I can set a "nice name" for using it for commands like "Ambient Light" instead of "dLedStripe_Swt"
> 
> Thanks again

OK so just to clarify,
i tried to change metadata through Sepia for a GUI-Generated item
it worked!

So i think this issue can be closed

Everything works as expected and metadata for Items created in *.items files have to be set there (as all other metadata too)

At the beginning i just was not sure if Sepia is chaning OHs items Metadata or does store its own metadata

Thanks for your support!
And yes, token also works for me ;-)

**Comment by SHU-red** [comment]
Just as a reminder
I think this topic is solved

Great job @fquirin
